modules/core/include/opencv2/core/hal/intrin_sse.hpp

```diff
@@ -1570,6 +1570,36 @@ inline void v_load_deinterleave(const unsigned* ptr, v_uint32x4& a, v_uint32x4&
     v_transpose4x4(u0, u1, u2, u3, a, b, c, d);
 }
 
+inline void v_load_deinterleave(const float* ptr, v_float32x4& a, v_float32x4& b, v_float32x4& c)
+{
+    __m128 t0 = _mm_loadu_ps(ptr + 0);
+    __m128 t1 = _mm_loadu_ps(ptr + 4);
+    __m128 t2 = _mm_loadu_ps(ptr + 8);
+    __m128 at12 = _mm_shuffle_ps(t1, t2, 0x12);
+    a.val = _mm_shuffle_ps(t0, at12, 0x8c);
+    __m128 bt01 = _mm_shuffle_ps(t0, t1, 0x01);
+    __m128 bt12 = _mm_shuffle_ps(t1, t2, 0x23);
+    b.val = _mm_shuffle_ps(bt01, bt12, 0x88);
+    __m128 ct01 = _mm_shuffle_ps(t0, t1, 0x12);
+    c.val = _mm_shuffle_ps(ct01, t2, 0xc8);
```

Comment:
This implementation is quite different from `v_load_deinterleave()` for other 32-bit types (see `v_uint32x4` at line 1548)

Perhaps it is better to reuse exited function and load as `v_unit32x4` + `v_reinterpret_as_f32()`

Reply:
Just a reminder, there are these choices: 
 * UNPCKLPS / UNPCKHPS ( ```_mm_unpacklo_ps``` , ```_mm_unpackhi_ps``` )
 * The confusion coming from the difference between SHUFPS and PSHUFD can be resolved, if you pass in the same register into the two inputs of SHUFPS, i.e. ```_mm_shuffle_ps( same_reg, same_reg, shuffle_code )```
 * To emulate _mm_unpacklo_epi64 with 32-bit floating point lanes, you can define an inline function with hard-coded SHUFPS shuffle_code. To make it easier to remember, the inline function's name can spell out the shuffle code's four digits in base-four. (On x86 intrinsics headers, there is a macro, ```_MM_SHUFFLE``` (all capitalized), which converts the four permutation digits from base-four, into the 8-bit compile-time constant.)

Reply:
Maybe your code is not faster than my code.

My code use _mm_shuffle_ps 7 times.
So latency is 7 and throughput is 7.

Your code (v_uint32x4 and v_reinterpret_as_f32) use
_mm_unpacklo_epi32 6times, _mm_unpackhi_epi64 6times.
So latency is 12 and throughput is 12.
Only IvyBridge, latency is 12 and throughput is 6.

Do I have to change my code?

Reply:
@tomoaki0705 

>Changing the implementation of existing universal intrinsic affects much larger/wider than
> you think, (i.e. can you prove that your modification doesn't degrade any single function
> in entire OpenCV ?)

I can't prove that my modification doesn't degrade.
But wasted unpack code is not faster than my code.

>You're not the only one who reads this code. It's easy to read by keeping consistency
> among other 32bit deinterleave.

I don't think so. Many unpack code for 3 plane is hard to read.
But if you think easy to read many unpack code for 3 plane, I will change my code.

>About the performance, talking ONLY on latency is bit dangerous.
>Modern CPU has super scalar and out-of-order execution.
>This means latency could be hidden by other instruction on run-time.
>It's better to talk in seconds, not clocks.

Latency could be hidden by other instruction only at IvyBridge.
But latency cound not be hidden at other CPU. (ex. Haswell, Broadwell, Skylake)  
Please see intel intrinsics guide.

>Also, "average" varies on test case, so I recommend to use "median" value instead of average.
>Average could be affected by very slow/fast one time exceptional result, and median value is >much worth to trust.
>(That's one of the reason why OpenCV performance test measures the median as long as the >average)

1. I measure clocks 10 times.
2. I pickup 5 data at 1's result.
3. I calculate average 2's data.
So my measurement result is not include exceptional result.

>More than that, don't use magical 0x12 values for shuffle.
>Use _MM_SHUFFLE instead.

I don't think that Engineers that can write simd code are difficult to read shuffle's magic code.
But if you think hard to read, I will change my code.
I'm sorry to trouble to you, please write code using _MM_SHUFFLE.

Reply:
About the throughput, your're right. it's my mistake.

> But wasted unpack code is not faster than my code.

In another word, then why don't you replace entire deinterleave/interleave by your approach ?
Then it'll keep the implementation in one place, and still it will speed up OpenCV, right ?
My concern is NOT ONLY about hard to read .  It's that having two different implementation doing same process is non-sense.

>  I measure clocks 10 times.
>   I pickup 5 data at 1's result.
>   I calculate average 2's data.
>   So my measurement result is not include exceptional result.

This is insane,,,,, which 5 data are you picking ?
More than that, what's the point doing everything in manual, when you already have implemented performance tests ?

> I don't think that Engineers that can write simd code are difficult to read shuffle's magic code.
But if you think hard to read, I will change my code.
> I'm sorry to trouble to you, please write code using _MM_SHUFFLE.

Look, ```_MM_SHUFFLE``` is used [here](https://github.com/opencv/opencv/blob/875b4e212a186afd9eb03af8b39d52b03a330c8d/modules/core/include/opencv2/core/hal/intrin_sse.hpp#L520), [here](https://github.com/opencv/opencv/blob/875b4e212a186afd9eb03af8b39d52b03a330c8d/modules/core/src/dxt.cpp#L391) and [here](https://github.com/opencv/opencv/blob/875b4e212a186afd9eb03af8b39d52b03a330c8d/modules/core/src/lapack.cpp#L959).

Reply:
>In another word, then why don't you replace entire deinterleave/interleave by your approach ?
>Then it'll keep the implementation in one place, and still it will speed up OpenCV, right ?
>My concern is NOT ONLY about hard to read . It's that having two different implementation
> doing same process is non-sense.

I want alalek to decide the policy.
I say that my code is faster. Because I don't want to change my code.
Then if alalek say that you should change the code, I intend to change my code.
If alalek say that I adopt your code, I intend to keep my code.
I think that replacement for entire deinterleave/interleave code should be done after his decision.

> This is insane,,,,, which 5 data are you picking ?
>More than that, what's the point doing everything in manual, when you already have >implemented performance tests ?

Originally, don't you worry about that my measurement result include outlier?
I say my result doesn't include outlier.
Is it not enough that I measured the speed the before/after code at same condition?

>Look, _MM_SHUFFLE is used here, here and here.

I did't find the manual like "intel intrinsics guide".
So if you know the url for manual listed _MM_SHUFFLE, I want to be taught the url.
I am sorry that I have troubled you.

Reply:
The policy is not mine. There is just a logic:
- we should avoid providing code duplication, or different implementations for exact same functionality (in this case it is function for deinterleaved loading of 32-bit values).
- it is better to reuse OpenCV performance tests system. If you want to propose alternative implementation, then we should be able to check regressions on different platforms.